packages/core/crates/core-network/src/messaging.rs

```diff
@@ -0,0 +1,130 @@
+use core_crypto::derivation::derive_ping_pong;
+use serde::{Deserialize, Serialize};
+use utils_types::traits::{BinarySerializable};
+
+use crate::errors::Result;
+
+/// Implements the Control Message sub-protocol, which currently consists of Ping/Pong
+/// messaging for the HOPR protocol.
+#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub enum ControlMessage {
+    /// Ping challenge
+    Ping(PingMessage),
+    /// Pong response to a Ping
+    Pong(PingMessage),
+}
+
+impl ControlMessage {
+    /// Creates ping challenge message
+    pub fn generate_ping_request() -> Self {
+        let mut ping = PingMessage::default();
+        ping.nonce.copy_from_slice(&derive_ping_pong(None));
+        Self::Ping(ping)
+    }
+
+    /// Given the received ping challenge, generates an appropriate response to the challenge.
+    pub fn generate_pong_response(request: &ControlMessage) -> Result<Self> {
+        match request {
+            ControlMessage::Ping(ping) => {
+                let mut pong = PingMessage::default();
+                pong.nonce.copy_from_slice(&derive_ping_pong(Some(ping.nonce())));
+                Ok(Self::Pong(pong))
+            }
+            ControlMessage::Pong(_) => Err(MessagingError("invalid ping message".into())),
+        }
+    }
+
+    /// Given the original ping challenge message, verifies that the received pong response is valid
+    /// according to the challenge.
+    pub fn validate_pong_response(request: &ControlMessage, response: &ControlMessage) -> Result<()> {
+        if let Self::Pong(expected_pong) = Self::generate_pong_response(request).unwrap() {
+            match response {
+                ControlMessage::Pong(received_pong) => match expected_pong.nonce.eq(&received_pong.nonce) {
+                    true => Ok(()),
+                    false => Err(MessagingError("pong response does not match the challenge".into())),
+                },
+                ControlMessage::Ping(_) => Err(MessagingError("invalid pong response".into())),
+            }
+        } else {
+            Err(MessagingError("request is not a valid ping message".into()))
+        }
+    }
+
+    /// Convenience method to de-structure the ping message payload, if this
+    /// instance is a Ping or Pong.
+    pub fn get_ping_message(&self) -> Result<&PingMessage> {
+        match self {
+            ControlMessage::Ping(m) | ControlMessage::Pong(m) => Ok(m),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Default, Serialize, Deserialize)]
+pub struct PingMessage {
+    nonce: [u8; core_crypto::parameters::PING_PONG_NONCE_SIZE],
+}
+
+impl PingMessage {
+    /// Gets the challenge or response in this ping/pong message.
+    pub fn nonce(&self) -> &[u8] {
+        &self.nonce
+    }
+}
+
+#[cfg(not(feature = "compat-ping"))]
+impl utils_types::traits::AutoBinarySerializable<'_> for PingMessage {
+    const SIZE: usize = core_crypto::parameters::PING_PONG_NONCE_SIZE;
```

Comment:
This may not be needed, if the AutoBinaryTrait is derived from the serializable BinarySerializable.